url9ecf1e8297_nhamidi_ninja_tgz-pJ8-net_fec_openrq_test_TestRunnerJ8/src/net/fec/openrq/core/ArraySourceBlockEncoder.java

```diff
@@ -0,0 +1,318 @@
+
+        checkRepairSymbolESI(esi);
+        checkNumRepairSymbols(esi, numSymbols);
+
+        // check if we've got the intermediate symbols already
+        if (intermediateSymbols == null) {
+            intermediateSymbols = generateIntermediateSymbols();
+        }
+
+        // generate FEC Payload ID
+        FECPayloadID fpid = FECPayloadID.makeFECPayloadID(sbn, esi, fecParams);
+
+        // generate repair symbols
+        final List<ByteBuffer> bufs = new ArrayList<>();
+        int isi = esi + (Kprime - K);
+
+        for (int i = 0; i < numSymbols; i++, isi++) {
+
+            byte[] enc_data = LinearSystem.enc(Kprime, intermediateSymbols, new Tuple(Kprime, isi),
+                fecParams.symbolSize());
+
+            // generate repair symbol // TODO should we store the repair symbols generated?
+            EncodingSymbol repairSymbol = EncodingSymbol.newRepairSymbol(fpid, enc_data);
+
+            bufs.add(repairSymbol.transportData());
+        }
+
+        // return the repair packet
+        return (new RepairPacket(fpid, ImmutableList.copy(bufs)));
+    }
+
+    private void checkSourceSymbolESI(int esi) {
+
+        if (esi < 0 || esi >= K) {
+            throw new IllegalArgumentException("invalid encoding symbol ID");
+        }
+    }
+
+    // requires valid ESI
+    private final void checkNumSourceSymbols(int esi, int numSymbols) {
+
+        if (numSymbols < 1 || numSymbols > K - esi) {
+            throw new IllegalArgumentException("invalid number of symbols");
+        }
+    }
+
+    private void checkRepairSymbolESI(int esi) {
+
+        if (esi < K || esi > ParameterChecker.maxEncodingSymbolID()) {
+            throw new IllegalArgumentException("invalid encoding symbol ID");
+        }
+    }
+
+    // requires valid ESI
+    private final void checkNumRepairSymbols(int esi, int numSymbols) {
+
+        if (numSymbols < 1 || numSymbols > (1 + ParameterChecker.maxEncodingSymbolID() - esi)) {
+            throw new IllegalArgumentException("invalid number of symbols");
+        }
+    }
+
+    private byte[] generateIntermediateSymbols() {
+
+        // source block's parameters
+        int Ki = SystematicIndices.getKIndex(Kprime);
+        int S = SystematicIndices.S(Ki);
+        int H = SystematicIndices.H(Ki);
+        int L = Kprime + S + H;
+        int T = fecParams.symbolSize();
+
+        // generate LxL Constraint Matrix
+        byte[][] constraint_matrix = LinearSystem.generateConstraintMatrix(Kprime, T); // A
+
+        // allocate and initialize vector D
+        byte[][] D = new byte[L][T];
+        for (int row = S + H, index = 0; row < K + S + H; row++, index++)
+            D[row] = sourceSymbols[index].data();
+
+        // solve system of equations
+        byte[] C = null;
+        try {
+            C = LinearSystem.PInactivationDecoding(constraint_matrix, D, T, Kprime);
+        }
+        catch (SingularMatrixException e) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: While wrapping the caught exception into a custom one, information about the caught exception is being lost, including information about the stack trace of the exception.

**Fix**: If the caught exception object does not contain sensitive information, consider passing it as the &#34;rootCause&#34; or inner exception parameter to the constructor of the new exception before throwing the new exception. (Note that not all exception constructors support inner exceptions. Use a wrapper exception that supports inner exceptions.) 
[Learn more](https://www.ibm.com/support/pages/best-practice-catching-and-re-throwing-java-exceptions)